diff --git a/cunit/test_utils.c b/cunit/test_utils.c
index 31fbdf2..9ccdcd4 100644
--- a/cunit/test_utils.c
+++ b/cunit/test_utils.c
@@ -206,7 +206,7 @@ void passphrase_read_prompts_to_tty()
 			close(STDOUT_FILENO);
 			close(STDERR_FILENO);
 			close(masterfd);
-			freerdp_passphrase_read("Password: ", buffer, password_size);
+			freerdp_passphrase_read("Password: ", buffer, password_size, false);
 			close(slavefd);
 			exit(EXIT_SUCCESS);
 		}
@@ -266,7 +266,7 @@ void passphrase_read_reads_from_tty()
 			close(STDERR_FILENO);
 			close(masterfd);
 			close(pipe_ends[0]);
-			freerdp_passphrase_read("Password: ", buffer, password_size);
+			freerdp_passphrase_read("Password: ", buffer, password_size, false);
 			write(pipe_ends[1], buffer, password_size);
 			close(slavefd);
 			close(pipe_ends[1]);
@@ -342,7 +342,7 @@ void passphrase_read_turns_off_echo_during_read()
 			close(STDERR_FILENO);
 			close(masterfd);
 			close(slavefd);
-			freerdp_passphrase_read("Password: ", buffer, password_size);
+			freerdp_passphrase_read("Password: ", buffer, password_size, false);
 			close(child_slavefd);
 			exit(EXIT_SUCCESS);
 		}
@@ -416,7 +416,7 @@ void passphrase_read_resets_terminal_after_read()
 			close(STDERR_FILENO);
 			close(masterfd);
 			close(slavefd);
-			freerdp_passphrase_read("Password: ", buffer, password_size);
+			freerdp_passphrase_read("Password: ", buffer, password_size, false);
 			close(child_slavefd);
 			exit(EXIT_SUCCESS);
 		}
@@ -490,7 +490,7 @@ void passphrase_read_turns_on_newline_echo_during_read()
 			close(STDERR_FILENO);
 			close(masterfd);
 			close(slavefd);
-			freerdp_passphrase_read("Password: ", buffer, password_size);
+			freerdp_passphrase_read("Password: ", buffer, password_size, false);
 			close(child_slavefd);
 			exit(EXIT_SUCCESS);
 		}
@@ -542,7 +542,7 @@ void passphrase_read_prompts_to_stderr_when_no_tty()
 
 			dup2(stdin_pipe[0], STDIN_FILENO);
 			dup2(stderr_pipe[1], STDERR_FILENO);
-			freerdp_passphrase_read("Password: ", buffer, password_size);
+			freerdp_passphrase_read("Password: ", buffer, password_size, false);
 			exit(EXIT_SUCCESS);
 		}
 	}
@@ -594,7 +594,7 @@ void passphrase_read_reads_from_stdin_when_no_tty()
 
 			dup2(stdin_pipe[0], STDIN_FILENO);
 			dup2(stderr_pipe[1], STDERR_FILENO);
-			freerdp_passphrase_read("Password: ", buffer, password_size);
+			freerdp_passphrase_read("Password: ", buffer, password_size, false);
 			write(result_pipe[1], buffer, strlen(buffer) + (size_t) 1);
 			exit(EXIT_SUCCESS);
 		}
